### PR TITLE
Truly Dynamic Event

### PR DESCRIPTION
**Pull request by cshung**
This PR is currently in a prototype state. (It is easier to review by commits)

This change:

1. Upgrade TraceEvent dependencies from 3.1.7 to 3.1.9, which also require upgrading the project to 8.0, so that we can
2. Allow using custom TraceEvent library, built some support for using it, to
3. Parse DynamicEvent and allow them to be accessed in the Notebooks.

And some other cleanups

1. Run dotnet-format on the source code
2. Ignore xlf files
3. Deleted unwanted (and failing) outputs from Notebook examples.

The dynamic event consumption part can only work with https://github.com/microsoft/perfview/pull/2051.

### CODE REVIEW

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            else
+            {
+                result = null;
+                return false;
+            }
+        }
+
+        public override string ToString()
+        {
+            return "I am " + this.rawDynamic.Name + " with these fields: \n" + string.Join("\n", this.fieldValues.Select(kvp => kvp.Key + "->" + kvp.Value));
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:124]: Is the "I am" necessary here? Just returning the fields should be sufficient.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            {
+                string name = binder.Name;
+                if (index.TryGetValue(name, out var indexValue))
+                {
+                    // TODO, andrewau, handle the case where we have multiple dynamic event of the same name in a single GC.
+                    //
+                    // That should be an array of these dynamic events
+                    // Somewhere the schema should tell us if this is expected or not so not to generate wrong stuff just because the data is corrupted.
+                    // 
+                    result = new DynamicEvent(indexValue[0], DynamicEventSchema.DynamicEventSchemas.Single(s => string.Equals(s.DynamicEventName, binder.Name)).Fields);
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:58]: Single will except if a unique or no item is found. Do we really want to except here without context? I would do a `SingleOrDefault` or try-catch the error and then contextualize it with other details vs. a generic call stack.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs (hunk 1 above)
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:124]: This is really just telling you "I am a prototype 😄".

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+            {
+                result = null;
+                return false;
+            }
+        }
+    }
+
+    internal class DynamicEvent : DynamicObject
+    {
+        private RawDynamic rawDynamic;
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:77]: These two fields can be made ``readonly``.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+#if CUSTOM_TRACE_EVENT
+
+using Microsoft.Diagnostics.Tracing.Analysis.GC;
+using Microsoft.Diagnostics.Tracing.Parsers.GCDynamic;
+using System.Diagnostics;
+using System.Dynamic;
+
+namespace GC.Analysis.API.DynamicEvents
+{
+
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:11]: NIT: Space.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/GC.Analysis.API.csproj [hunk 5]
```diff
@@ -2,21 +2,32 @@
...
     <GenerateTargetFrameworkAttribute>false</GenerateTargetFrameworkAttribute> 
   </PropertyGroup>
 
+  <Import Project="../Versions.props" />
+
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.Html.Abstractions" Version="2.2.0" />
     <PackageReference Include="Microsoft.Data.Analysis" Version="0.19.1" />
-    <PackageReference Include="Microsoft.Diagnostics.Tracing.TraceEvent" Version="3.1.7" />
+    <PackageReference Condition="'$(CustomTraceEvent)' != 'true'" Include="Microsoft.Diagnostics.Tracing.TraceEvent" Version="3.1.9" />
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/GC.Analysis.API.csproj:17]: Was this condition intentional?

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/GC.Analysis.API.csproj [hunk 6]
```diff
@@ -2,21 +2,32 @@
...
     <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
     <PackageReference Include="XPlot.Plotly" Version="4.0.6" />
     <PackageReference Include="YamlDotNet" Version="11.2.1" />
   </ItemGroup>
+  
+  <ItemGroup Condition="'$(CustomTraceEvent)' == 'true'">
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/GC.Analysis.API.csproj:23]: Why were these dependencies necessary? These are from the PerfView code and shouldn't be included here.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs (hunk 2 above)
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:58]: Yes, I am still trying to figure out the pattern here, something along the line of having the schema to provide some constraints, so this can do the proper handling.

Something like:

```c#
enum EventConstraint
{
  ExactlyOne,
  AtMostOne,
  OneOrMore,
  ZeroOrMore
}
```

File: src/benchmarks/gc/GC.Infrastructure/Notebooks/Examples/GCAnalysisExamples.ipynb [hunk 7]
```diff
@@ -19,19 +19,9 @@
...
-      ]
-     },
-     "metadata": {},
-     "output_type": "display_data"
-    }
-   ],
+   "outputs": [],
    "source": [
-    "#r \"nuget: Microsoft.Diagnostics.Tracing.TraceEvent, 3.1.7\"\n",
+    "#r \"nuget: Microsoft.Diagnostics.Tracing.TraceEvent, 3.1.9\"\n",
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/Notebooks/Examples/GCAnalysisExamples.ipynb:24]: Why weren't the other notebooks updated with the latest version of TraceEvent?

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/GC.Analysis.API.csproj (hunk 5 above)
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/GC.Analysis.API.csproj:17]: Yes, it is a feature of this change.

If you look at `Versions.props`, there is a switch where you can flip back and forth using either a Nuget package or a locally build TraceEvent.

This is the enabler for us to quickly iterate different versions of TraceEvent.

File: src/benchmarks/gc/GC.Infrastructure/Versions.props [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project> 
+  <PropertyGroup>
+    <!-- To use a custom TraceEvent library, specify CustomTraceEvent to true and the PerfViewPath below -->
+    <CustomTraceEvent>false</CustomTraceEvent>
+    <PerfViewPath>$(MSBuildProjectDirectory)\..\..\..\..\..\..\PerfView</PerfViewPath>
+    <DefineConstants Condition="'$(CustomTraceEvent)' == 'true'">CUSTOM_TRACE_EVENT;$(DefineConstants)</DefineConstants>
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/Versions.props:7]: Is this necessary or can be removed? We hadn't needed it before.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/GC.Analysis.API.csproj (hunk 6 above)
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/GC.Analysis.API.csproj:23]: It won't compile otherwise because if `CustomTraceEvent` is set to true, the nuget package above is skipped.

File: src/benchmarks/gc/GC.Infrastructure/Versions.props (hunk 8 above)
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/Versions.props:7]: The whole point of this change is to make it possible to use a local version where we never able to before.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+        public static dynamic DynamicEvents(this TraceGC traceGC)
+        {
+            return new DynamicIndex(traceGC.dynamicEventIndex);
+        }
+    }
+
+    // TODO, AndrewAu, this could be compiled for better performance, right now we are doing linear lookups
+    // We can do validation so that the input make sense (i.e. no unexpected types, no duplicate event/field), etc
+    //
+    public class DynamicEventSchema
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:23]: Can we seal this class?

Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:23]: Yes, we can.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        public DynamicEvent(RawDynamic rawDynamic, List<KeyValuePair<string, Type>> fields)
+        {
+            this.rawDynamic = rawDynamic;
+            this.fieldValues = new Dictionary<string, object>();
+            int offset = 0;
+            foreach (KeyValuePair<string, Type> field in fields)
+            {
+                object value = null;
+                // TODO, AndrewAu, all types that we envision this will be needed
+                if (field.Value == typeof(ushort))
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:89]: What about arrays of numerics? Should we handle those in a first pass?

File: src/benchmarks/gc/GC.Infrastructure/GC.Infrastructure.Core/GC.Infrastructure.Core.csproj [hunk 11]
```diff
@@ -23,5 +25,14 @@
   <ItemGroup>
     <ProjectReference Include="..\GC.Analysis.API\GC.Analysis.API.csproj" />
   </ItemGroup>
+  
+  <ItemGroup Condition="'$(CustomTraceEvent)' == 'true'">
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Infrastructure.Core/GC.Infrastructure.Core.csproj:29]: The unit tests make no use of the dynamic events, why is this needed here?

Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Infrastructure.Core/GC.Infrastructure.Core.csproj:29]: I don't understand fully either, but without directly referencing this, the project won't build because somewhere in the result comparison we accessed the properties of `GCStats` that was exposed on the API.

If the API perform the access and return raw values instead this wouldn't be necessary.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs (hunk 10 above)
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:89]: I don't think we can emit array of numbers right now in dynamic event, can we?

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        private Dictionary<string, List<RawDynamic>> index;
+
+        public DynamicIndex(Dictionary<string, List<RawDynamic>> index)
+        {
+            this.index = index;
+        }
+
+        public override bool TryGetMember(GetMemberBinder binder, out object result)
+        {
+            if (DynamicEventSchema.DynamicEventSchemas.Any(s => string.Equals(s.DynamicEventName, binder.Name)))
```
[the remark this answers is not in the document]
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:48]: I don't know if I want case insensitive though, doesn't seem like a big deal of convenience.

File: src/benchmarks/gc/GC.Infrastructure/Notebooks/Examples/GCAnalysisExamples.ipynb (hunk 7 above)
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/Notebooks/Examples/GCAnalysisExamples.ipynb:24]: Ooops, they should be, I just missed.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs (hunk 2 above)
Comment (cshung) [reply src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:58]: The "More" option is what I really need to think more about - imagine we emit two dynamic events of the same type before and after a GC. We probably would like to be able to get an array of them. Right now we can't.

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+        private static ConditionalWeakTable<TraceGC, DynamicIndex> dynamicEventsCache = new ConditionalWeakTable<TraceGC, DynamicIndex>();
+
+        public static dynamic DynamicEvents(this TraceGC traceGC)
+        {
+            DynamicIndex dynamicIndex;
+            if (!dynamicEventsCache.TryGetValue(traceGC, out dynamicIndex))
+            {
+                dynamicIndex = new DynamicIndex(traceGC.DynamicEvents);
+                dynamicEventsCache.Add(traceGC, dynamicIndex);
+            }
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:26]: Nit: add a line here?

File: src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+using System.Runtime.CompilerServices;
+using System.Text;
+
+[assembly: InternalsVisibleTo("GC.Analysis.API.UnitTests")]
+
+namespace GC.Analysis.API.DynamicEvents
+{
+    public static class TraceGCExtensions
+    {
+        private static ConditionalWeakTable<TraceGC, DynamicIndex> dynamicEventsCache = new ConditionalWeakTable<TraceGC, DynamicIndex>();
```
Comment (mrsharm) [inline src/benchmarks/gc/GC.Infrastructure/GC.Analysis.API/DynamicEvents/DynamicEvents.cs:17]: Can this be a readonly field?

**Review by mrsharm — DISMISSED** [review]
LGTM - just some nits here and there but we tested it out and it all works.

**Review by mrsharm — APPROVED** [review]
LGTM!

**Comment by mrsharm** [comment]
@cincuranet - could we please get this PR merged, seems like the Build Analysis failures are unrelated:

❌The pipeline is not valid. Job windows_22H2_x64_scenarios__Open: Step AzureCLI input connectedServiceNameARM references service connection .NET Performance (790c4451-dad9-4fda-af8b-10bd9ca328fa) which could not be found. The service connection does not exist, has been disabled or has not been authorized for use. For authorization details, refer to https://aka.ms/yamlauthz. Job ubuntu_2204_x64_scenarios__Open: Step AzureCLI input connectedServiceNameARM references service connection .NET Performance (790c4451-dad9-4fda-af8b-10bd9ca328fa) which could not be found. The service connection does not exist, has been disabled or has not been authorized for use. For authorization details, refer to https://aka.ms/yamlauthz.

Thanks!

**Comment by cincuranet** [comment]
@LoopedBard3 @DrewScoggins @caaavik-msft @sblom PTAL

**Comment by LoopedBard3** [comment]
Error was fixed earlier this week and this PR is approved by correct people, merging 👍.
